### PR TITLE
Feature/jra 26 android register user

### PR DESCRIPTION
**Pull request by Rodrigo-Schar**
- Added Register User
- Added Firebase Auth
- Added Email validation
- added hook and fix request changes
- added date to Double, observers with viewModel
- merge and fix model
- Added UnitTest for validateFields

### CODE REVIEW

File: app/src/main/java/com/jalasoft/routesapp/data/remote/managers/FirebaseManager.kt [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+                successListener(documentReference.id)
+            }.addOnFailureListener { exception ->
+                errorListener(exception.message.toString())
+            }
+        }
+
+        fun getUsersByParameter(collection: FirebaseCollections, field: String, parameter: String, successListener: (MutableList<User>) -> Unit, errorListener: (String) -> Unit) {
+
+            db.collection(collection.toString()).whereEqualTo(field, parameter).get().addOnSuccessListener { documents ->
+                //var users = mutableListOf<User>()
```
Comment (jmacboy) [inline app/src/main/java/com/jalasoft/routesapp/data/remote/managers/FirebaseManager.kt]: Remove unnecessary comments

File: app/src/main/java/com/jalasoft/routesapp/data/remote/managers/UserManager.kt [hunk 2]
```diff
@@ -0,0 +1,39 @@
+package com.jalasoft.routesapp.data.remote.managers
+
+import com.jalasoft.routesapp.data.model.remote.User
+import com.jalasoft.routesapp.util.helpers.DateHelper
+import com.jalasoft.routesapp.util.helpers.FirebaseCollections
+import com.jalasoft.routesapp.util.helpers.UserType
+import com.jalasoft.routesapp.util.helpers.UserTypeLogin
+
+class UserManager {
+    object Singleton {
```
Comment (jmacboy) [inline app/src/main/java/com/jalasoft/routesapp/data/remote/managers/UserManager.kt]: There’s no need to declare a Singleton object, if you’re just going to use static methods, just change class UserManager for object UserManager

File: app/src/main/java/com/jalasoft/routesapp/ui/auth/login/view/LoginFragment.kt [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+    }
+
+    override fun onCreateView(
+        inflater: LayoutInflater, container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View? {
+        // Inflate the layout for this fragment
```
Comment (jmacboy) [inline app/src/main/java/com/jalasoft/routesapp/ui/auth/login/view/LoginFragment.kt]: Remove unnecessary comments

File: app/src/main/java/com/jalasoft/routesapp/ui/auth/registerUser/view/RegisterUserFragment.kt [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+    private fun addUser() {
+        if (validateFields()) {
+            showProgress(true)
+            val name = binding.etRegName.text.toString()
+            val email = binding.etRegEmail.text.toString()
+            val password = binding.etRegPassword.text.toString()
+
+            viewModel.registerUserAuth(name, email, password, { _ ->
+                binding.txtRegError.isVisible = false
```
Comment (jmacboy) [inline app/src/main/java/com/jalasoft/routesapp/ui/auth/registerUser/view/RegisterUserFragment.kt]: This is wrong, success and failure should be handled on the viewmodel and update variables, here you just need to listen to variable changes

File: app/src/main/java/com/jalasoft/routesapp/ui/auth/registerUser/view/RegisterUserFragment.kt [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+            }, { error ->
+                showProgress(false)
+                binding.txtRegError.isVisible = true
+                binding.txtRegError.setText(error)
+            })
+        }
+    }
+
+    private fun validateFields(): Boolean {
+        var isValid = true
```
Comment (jmacboy) [inline app/src/main/java/com/jalasoft/routesapp/ui/auth/registerUser/view/RegisterUserFragment.kt]: Method is good but the logic should be moved to the viewmodel and update the variables in there. When the variables are updated refresh the fragment here

File: app/src/main/java/com/jalasoft/routesapp/ui/auth/registerUser/viewModel/RegisterUserViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+import com.jalasoft.routesapp.data.model.remote.User
+import com.jalasoft.routesapp.data.remote.managers.UserManager
+
+class RegisterUserViewModel: ViewModel() {
+
+    fun registerUserAuth(name: String, email: String, password: String, successListener: (String) -> Unit, errorListener: (String) -> Unit) {
+        if (validateEmail(email)) {
+            UserManager.Singleton.createUserAuth(email, password, { _ ->
+                registerUser(name, email, { success ->
+                    successListener(success)
```
Comment (jmacboy) [inline app/src/main/java/com/jalasoft/routesapp/ui/auth/registerUser/viewModel/RegisterUserViewModel.kt]: The logic should be handled here

File: app/src/main/res/layout/fragment_login.xml [hunk 7]
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="utf-8"?>
+<FrameLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    tools:context=".ui.auth.login.view.LoginFragment">
+
+    <!-- TODO: Update blank fragment layout -->
```
Comment (jmacboy) [inline app/src/main/res/layout/fragment_login.xml]: Remove unnecessary comments

File: app/src/main/res/values/themes.xml [hunk 8]
```diff
@@ -13,4 +13,18 @@
         <item name="android:statusBarColor">?attr/colorPrimaryVariant</item>
         <!-- Customize your theme here. -->
     </style>
+
+    <!-- Rounded Button. -->
+    <style name="btnPrimaryRounded" parent="Widget.MaterialComponents.Button.OutlinedButton">
+        <item name="shapeAppearanceOverlay">@style/MyShapeAppearance</item>
+    </style>
+    <style name="MyShapeAppearance">
```
Comment (jmacboy) [inline app/src/main/res/values/themes.xml]: Add a more descriptive name to this appearance

**Review by jmacboy — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/jalasoft/routesapp/ui/auth/registerUser/view/RegisterUserFragment.kt (hunk 5 above)
Comment (Rodrigo-Schar) [reply app/src/main/java/com/jalasoft/routesapp/ui/auth/registerUser/view/RegisterUserFragment.kt]: Done!

File: app/src/main/java/com/jalasoft/routesapp/ui/auth/registerUser/view/RegisterUserFragment.kt (hunk 4 above)
Comment (Rodrigo-Schar) [reply app/src/main/java/com/jalasoft/routesapp/ui/auth/registerUser/view/RegisterUserFragment.kt]: Done!

Comment (jmacboy) [reply app/src/main/java/com/jalasoft/routesapp/ui/auth/registerUser/view/RegisterUserFragment.kt]: This is still wrong, this logic should be handled on the viewmodel, and you should be observing variable changes on the Fragment. I don't see a place where this is happening, you´re trying to imitate that with the callback but it isn't that way.

imagen

Example from this link:
https://developer.android.com/topic/libraries/architecture/livedata#observe_livedata_objects

Comment (Rodrigo-Schar) [reply app/src/main/java/com/jalasoft/routesapp/ui/auth/registerUser/view/RegisterUserFragment.kt]: i think now it is good i add the observer for the viewModel you can check it

**Review by YawarValeriano — APPROVED** [review]
Seems good to me

File: app/src/main/java/com/jalasoft/routesapp/data/model/remote/User.kt [hunk 9]
```diff
@@ -0,0 +1,12 @@
+package com.jalasoft.routesapp.data.model.remote
+
+data class User(
+    val Id: String? = null,
```
Comment (YawarValeriano) [inline app/src/main/java/com/jalasoft/routesapp/data/model/remote/User.kt]: We should define naming variables form (camelCase, PascalCase)

**Review by YawarValeriano — CHANGES_REQUESTED** [review]

**Comment by YawarValeriano** [comment]
Also pull changes from develop to add de sonar gradle config for this pull request

**Review by YawarValeriano — APPROVED** [review]

File: app/src/main/java/com/jalasoft/routesapp/data/model/remote/User.kt [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+
+data class User(
+    val id: String? = null,
+    val name: String? = null,
+    val email: String? = null,
+    val phoneNumber: String? = null,
+    val type: Int? = null,
+    val typeLogin: Int? = null,
+    val createdAt: Double? = null,
+    val updatedAt: Double? = null
```
Comment (jmacboy) [inline app/src/main/java/com/jalasoft/routesapp/data/model/remote/User.kt:11]: Does the date has decimals? or why double instead of long

**Review by jmacboy — CHANGES_REQUESTED** [review]
the other stuff is ok now, just one question this time

**Comment by jmacboy** [comment]
Also I see 0 unit tests on this code @Rodrigo-Schar

File: app/src/main/java/com/jalasoft/routesapp/data/model/remote/User.kt (hunk 10 above)
Comment (Rodrigo-Schar) [reply app/src/main/java/com/jalasoft/routesapp/data/model/remote/User.kt:11]: because in ios it is saving with decimals so to don't get issues

**Comment by jmacboy** [comment]
just missing unit tests and you´re good to go

**Comment by Rodrigo-Schar** [comment]
i changed my validateFuntion to try to don't test the liveData, i tried with others options to test the liveData but nothing works , tomorrow Alvaro will show an example of how to test with mock so i'm not sure if i can pull this request with that two unitest that i added

**Review by jmacboy — APPROVED** [review]
I´m approving it by now, so you can merge, but you really need to fix the unit test issues

**Pull request merged**